### PR TITLE
debug: coredump: add new backend to store dump in flash, shell support and misc fixes

### PR DESCRIPTION
**Pull request by dcpleung**
This adds a new backend to use a flash partition to store coredump, and associated shell module to retrieve the stored coredump. 

This also fixes an issue in the Python script where a non-existent memory region would crash the script.

### CODE REVIEW

File: include/debug/coredump.h [hunk 1]
```diff
@@ -136,6 +170,30 @@ void z_coredump_buffer_output(uint8_t *buf, size_t buflen)
  * @param buflen Buffer length
  */
 
+/**
+ * @fn int z_coredump_query(enum z_coredump_query_id query_id, void *arg);
```
Comment (nashif) [inline include/debug/coredump.h:203]: are those public APIs? if yes, we should not be using z_, z_ is only for private APIs used within the subsystem itself, not applicaton or user facing.

Comment (dcpleung) [reply include/debug/coredump.h:203]: My thinking is that these are not to be used as-is, and will always be used under another layer of subsystem (e.g. telemetry). That's why these are not public facing APIs.

Comment (nashif) [reply include/debug/coredump.h:203]: ok, we should remove z_ prefix. z_ is basically a replacement for plain _ and signifies a private function within a subsystem. This is a public header and the APIs have documentation and will be used to add new backends and such, so it is not private and defines an interface. Even if not application facing.

Comment (dcpleung) [reply include/debug/coredump.h]: Removed the z_ prefix in public APIs.

**Review by nashif — APPROVED** [review]

File: subsys/debug/coredump/coredump_core.c [hunk 2]
```diff
@@ -158,3 +158,29 @@ void z_coredump_memory_dump(uintptr_t start_addr, uintptr_t end_addr)
...
+{
+	int ret;
+
+	if (backend_api->query == NULL) {
+		ret = -ENOTSUP;
+	} else {
+		ret = backend_api->query(query_id, arg);
+	}
+
+	return ret;
```
Comment (jenmwms) [inline subsys/debug/coredump/coredump_core.c:176]: just a thought - we could condense this, unless we see it scaling up to different conditional statements. 
`return (backend_api->query == NULL) ? (-ENOTSUP) : (backend_api->query(query_id, arg));`

File: subsys/debug/coredump/coredump_core.c [hunk 3]
```diff
@@ -158,3 +158,29 @@ void z_coredump_memory_dump(uintptr_t start_addr, uintptr_t end_addr)
...
+{
+	int ret;
+
+	if (backend_api->cmd == NULL) {
+		ret = -ENOTSUP;
+	} else {
+		ret = backend_api->cmd(cmd_id, arg);
+	}
+
+	return ret;
```
Comment (jenmwms) [inline subsys/debug/coredump/coredump_core.c:189]: similar comment here about condensing this part. not a blocker.

File: subsys/debug/coredump/coredump_core.c (hunk 2 above)
Comment (nashif) [reply subsys/debug/coredump/coredump_core.c:176]: the code as it is is easier to read, you can add comment, if it changes, you it is easier to review in form of a patch, and above all, there is no gain in making this a one liner.

Comment (jenmwms) [reply subsys/debug/coredump/coredump_core.c:176]: ok, understood. it was just a thought.

File: subsys/debug/coredump/coredump_core.c (hunk 3 above)
Comment (jenmwms) [reply subsys/debug/coredump/coredump_core.c:189]: nvm see above reply

**Review by jenmwms — APPROVED** [review]

**Pull request merged**